src/OrchardCore/OrchardCore.Abstractions/Modules/ModularApplicationContext.cs

```diff
@@ -36,21 +36,24 @@ private void EnsureInitialized()
         }
     }
 
-    private ConcurrentBag<Module> GetModules()
+    private List<Module> GetModules()
     {
-        var modules = new ConcurrentBag<Module>
-        {
-            new(_environment.ApplicationName, true),
-        };
-
+        var modules = new List<Module> { new(_environment.ApplicationName, true) };
         var names = _moduleNamesProviders
             .SelectMany(p => p.GetModuleNames())
             .Where(n => n != _environment.ApplicationName)
-            .Distinct();
+            .Distinct()
+            .ToList();
+
+        object lockObj = new object();
 
         Parallel.ForEach(names, new ParallelOptions { MaxDegreeOfParallelism = 8 }, (name) =>
         {
-            modules.Add(new Module(name, false));
+            var module = new Module(name, false);
+            lock (lockObj) // Ensures thread-safe addition
```

Comment:
There is nothing to optimize in the class by changing the structure. This parallel loop is completely useless, it should be a straight loop or LINQ query.

If there was something to be done in parallel it would be the calls to `GetModuleNames()`, but even that is useless because of how it's currently implemented in the class, the costly logic being in the constructor.

So one thing to try is to move the logic of loading assemblies into `GetModuleNames()` and then parallelizing these calls, not the code adding to the list.

Finally add some log at the beginning and end of this method, with and without the code changes to show how much is gained. No need to parallelize if there are no gains.